sycl/include/sycl/ext/intel/experimental/pipes.hpp

```diff
@@ -11,29 +11,110 @@
 #include "fpga_utils.hpp"
 #include <CL/__spirv/spirv_ops.hpp>
 #include <CL/__spirv/spirv_types.hpp>
+#include <sycl/context.hpp>
+#include <sycl/device.hpp>
+#include <sycl/ext/intel/experimental/pipe_properties.hpp>
 #include <sycl/ext/oneapi/properties/properties.hpp>
+#include <sycl/queue.hpp>
 #include <sycl/stl.hpp>
 #include <type_traits>
 
+#ifdef XPTI_ENABLE_INSTRUMENTATION
+#include <xpti/xpti_data_types.h>
+#include <xpti/xpti_trace_framework.hpp>
+#endif
+
 namespace sycl {
 __SYCL_INLINE_VER_NAMESPACE(_V1) {
-namespace ext::intel::experimental {
+namespace ext {
+namespace intel {
+namespace experimental {
+
+// A helper templateless base class to get the host_pipe name.
+class pipe_base {
+
+protected:
+  pipe_base();
+  ~pipe_base();
+
+  __SYCL_EXPORT static std::string get_pipe_name(const void *HostPipePtr);
+};
 
 template <class _name, class _dataT, int32_t _min_capacity = 0,
           class _propertiesT = decltype(oneapi::experimental::properties{}),
           class = void>
-class pipe {
-  static_assert(std::is_same_v<_propertiesT,
-                               decltype(oneapi::experimental::properties{})>,
-                "experimental pipe properties are not yet implemented");
-};
-
-template <class _name, class _dataT, int32_t _min_capacity, class _propertiesT>
-class pipe<_name, _dataT, _min_capacity, _propertiesT,
-           std::enable_if_t<std::is_same_v<
-               _propertiesT, decltype(oneapi::experimental::properties{})>>> {
+class pipe : public pipe_base {
 public:
+  struct
+#ifdef __SYCL_DEVICE_ONLY__
+      [[__sycl_detail__::add_ir_attributes_global_variable(
+          "sycl-host-pipe", "sycl-host-pipe-size", nullptr,
+          sizeof(_dataT))]] [[__sycl_detail__::sycl_type(host_pipe)]]
+#endif // __SYCL_DEVICE_ONLY___
+      ConstantPipeStorageExp
+#ifdef __SYCL_DEVICE_ONLY__
+      : ConstantPipeStorage
+#endif // __SYCL_DEVICE_ONLY___
+  {
+    int32_t _ReadyLatency;
+    int32_t _BitsPerSymbol;
+    bool _UsesValid;
+    bool _FirstSymInHighOrderBits;
+    protocol_name _Protocol;
+  };
+
   // Non-blocking pipes
+
+  // Host API
+  static _dataT read(queue &Q, bool &Success,
+                     memory_order Order = memory_order::seq_cst) {
+    const device Dev = Q.get_device();
+    bool IsPipeSupported =
+        Dev.has_extension("cl_intel_program_scope_host_pipe");
+    if (!IsPipeSupported) {
+      return _dataT();
+    }
+    _dataT Data;
+    void *DataPtr = &Data;
+    const void *HostPipePtr = &m_Storage;
+    const std::string PipeName = pipe_base::get_pipe_name(HostPipePtr);
+
+    event E = Q.submit([=](handler &CGH) {
+      CGH.ext_intel_read_host_pipe(PipeName, DataPtr, sizeof(_dataT) /* non-blocking */);
+    });
+    E.wait();
```

Comment:
What is the point of it being non-blocking if we immediately wait for it here?

Reply:
Nevermind. I misunderstood the point of the blocking behavior.

Reply:
@steffenlarsen 
This blocking parameter is used by the lower OpenCL function. If this is set to be non-blocking, the OpenCL function will not wait forever for the incoming data, and send event complete message right away regardless whether there is any data to read. If it's blocking, the opencl function will wait forever until there is data to read.